### PR TITLE
Make OCaml native debugging awesome

### PR DESCRIPTION
**Pull request by mshinwell**
This work lays the foundations for a good user experience when debugging OCaml code in a native platform debugger such as gdb or lldb.

This patch is a work in progress, needing a lot more effort, but I hope to be able to present it in a final form in time for 4.04.  The aim of showing it now is so that people who are interested can start getting a feel for the code.

This patch aims to provide in the debugger:
- naming of OCaml functions and values via their usual module paths, including submodules, e.g. Foo.Bar.myfunction---this will enable those names to be used for setting breakpoints, to be seen in backtraces, etc;
- visibility of function parameters, let-bound local variables and free variables of functions;
- printing of the values of such variables (by using the type information in the .cmt files).

Further work is also required to improve the precision of debugging events and the propagation of location information, some of which has been done, and will be presented separately.  That in particular is required so that stepping works more satisfactorily than it does at present.

To make use of this code it is necessary to have debugger-side support.  When this is in a more satisfactory state, hopefully in a couple of weeks' time, I will post a link here.  OCamlPro are working on lldb support although I do not know of their exact plans.  The debugger-side support will be in two parts:
1. a debugger-independent library, which will be able to be instantiated using [Obj] as the "debugger" to enable easy testing via introspection;
2. an extension hard-coded into gdb to understand OCaml.  This will probably use the debugger-independent library by loading it as a shared object combined with the OCaml runtime.

The patch has several main parts.  At present it is aimed specifically at use in Flambda mode.  Support for the old Closure mode may or may not be added later, depending on timescales.  Despite its size, the patch is for the most part orthogonal to existing source files, although there are various necessary changes scattered throughout the backend.
1. (At an early experimental stage.)  Modifications to Flambda so that values declared at toplevel are annotated with their full module paths.  (The compiler normally drops that information.)  So we should be able to type "break Foo.Bar.myfunction" in gdb.
2. (Likewise.) Modifications to Flambda and beyond so that values found constant during Flambda passes and lifted to toplevel, but not declared at toplevel, are still accessible using their original names in approximately their original scopes.  This is done via a "phantom let" construction, which is like a normal let-expression but only generates debugging information, not code.  It is the intention in the future to use "phantom let" to express other relationships between values, for example during pattern match compilation, to enable recovery of full or partial values in the debugger when they would otherwise be unavailable for printing.  For example we might know that x = Foo (a, b, c); x might be unavailable as might a, but b and c might be available.  In this case we should be able to print x = Foo (<unknown>, 42, 43) or whatever in the debugger.
3. (Ready for review.) A new forwards dataflow pass, Available_regs, which calculates register availability information.  There is an associated patch to the existing Liveness pass to ensure that liveness and availability information match up at allocation points, which may raise asynchronous exceptions.  The dataflow pass, which may be augmented in the future with intraprocedural dataflow analysis, tells us which registers or stack slots contain which named values when executing some particular part of the program.
4. (Ready for review.) A pass called Available_ranges that coalesces the output of Available_regs to obtain contiguous ranges of code where we know a certain value will be available at a certain location.
5. (Under development though mostly done.) Two libraries, Dwarf_low and Dwarf_high, which implement part of the DWARF-4 specification, in part using GADTs and polymorphic variants to ensure soundness of certain constructions.  There is not yet support for reading DWARF but this is likely to be added in the future.  Dwarf_low follows the specification and Dwarf_high provides a more user-friendly interface, in particular enabling DWARF debugging information trees to be constructed using simple parenting relationships.
6. (Under development.) One file, asmcomp/dwarf.ml, which takes the output of the register availability passes and further available range information deduced from "phantom let" expressions, and describes everything in DWARF.  Some of the constructs generated are very specific to the debugger support library.  (We may consider an approach to the DWARF standards committee in due course to try to get the standard extended to cover cases we need for OCaml without having to encode them in existing constructs.)

A guiding principle of this work is that the code to be debugged is intended to be the same as used in production systems.  In particular, there is no artificial extension of lifetimes of variables to enable more variables to be printed.

As above, there is significant work to do until this all works satisfactorily (and on all necessary platforms), but it is being actively worked on.  In any event, any first merged version is likely to only be a first attempt in some sense, this being a highly complicated problem.  For the moment I leave interested developers to browse the code, and will revert in due course when there is something working to try out.

### CODE REVIEW

**Comment by mshinwell** [comment]
I forgot to mention one thing: there is also a new abstraction layer (interface Asm_directives.S) which abstracts between the DSL-based x86 backend and other backends for the purposes of emission of assembly directives.  This part is mostly done for x86, and needs filling out for other textual-based backends (which should be straightforward).

**Comment by lijunsong** [comment]
By the way, does the compiler still support sparc platform? README says yes, reluctantly, but I couldn't build the compiler from source on sparc solaris.

**Comment by mshinwell** [comment]
@lijunsong : I think it officially still does, so please file a bug report at caml.inria.fr/mantis.

**Comment by xavierleroy** [comment]
SPARC support is in a zombie state: we have no SPARC/Solaris machine to run tests on, so eventually it will get out of sync with the OCaml code base and will have to be retired.  Unless someone with access to a SPARC/Solaris machine volunteers to run tests at regular intervals.

**Comment by lefessan** [comment]
@mshinwell OCamlPro's previous work on LLDB (`ocp-lldb` and other tools) was released last month: https://github.com/OCamlPro/typerex-lldb

We have now an intern working on improving the support for debugging in OCaml (improving locations) and porting your work to our version of OCaml with allocation identifiers (that we are planning to release at some point), that we are currently using with our memory profiler. The reason for that choice is that using .cmt files to recover type information is not very robust, in the sense that they might not be available at the time of debugging. On the contrary, our allocation identifiers are stored in the program with the corresponding type information, and `ocp-lldb` already uses them to print structured values.

**Comment by mshinwell** [comment]
I think in the long term, the right thing to do is probably to define some type system which is midway between the normal OCaml type system and the existing minimal type system of the heap.  It would provide sufficient information to describe the runtime representation of values without the abstraction barriers that we require in the normal type system.  This could then be propagated through the backend.  (It might also be usable for optimisation purposes.)  We would then describe as much as possible of those types using existing DWARF constructs, and propose to the DWARF standards committee extensions to the language to describe the remainder.  We have some people interested in working on this.

That aside for the moment, relying on .cmt files doesn't seem a big problem.  In particular, it is entirely consistent with the modern treatment of debugging information, namely that it exists in separate files.  (See /usr/lib/debug on Linux, for example.  This is also absolutely standard on Mac OS X, which uses .dSYM files.)

**Comment by mshinwell** [comment]
Typo correction: I meant "interprocedural dataflow" not "intraprocedural dataflow" (which is what we already do), above.

**Comment by lefessan** [comment]
I agree with your long term plan, but on the middle term, locids provide a nice workaround for the lack of expressivity of DWARF type systems. Moreover, locids also support being stored in external files, à la pbx.

**Comment by mshinwell** [comment]
Instructions to build:
- Install an OCaml compiler from this pull request.  It must be configured as PIC and with flambda, e.g.

```
./configure -prefix /path/to/ocaml/ -cc "gcc -fPIC" -aspp "gcc -c -fPIC" -flambda
```

Alternatively, add `git://github.com/mshinwell/opam-repo-dev` as an opam remote and use the `4.04.0+monda` switch.
- Install GDB from `https://github.com/mshinwell/gdb` branch `gdb-7.11`.  This is done as usual (`./configure --prefix=/path/to/gdb`, `make`, `make install`).
- Install `libmonda` from `https://github.com/mshinwell/libmonda`.  You need to edit the Makefile to point at your GDB source tree.  Further, when running the Makefile, the compiler you installed in step  1 needs to be on the path.

To compile a program for debugging, just build with `-g`.  This now implies `-bin-annot`.

To debug a program, run `gdb` with `LD_LIBRARY_PATH` including the directory where `libmonda_gdb.so` can be found.

This is still full of bugs, misfeatures and missing functionality.  Please don't report such things yet, but I am eager to here about potential use cases and desired features.

Here is an example of debugging a program, `small.ml`, that has been optimized by Flambda.

```
(gdb) b Small.My_submodule.f_sub 
Breakpoint 1 at 0x41ad70: file small.ml, line 2.
(gdb) r
Starting program: small

Breakpoint 1, Small.My_submodule.f_sub (x=42 : int) at small.ml:2
2     let f_sub x =
(gdb) list
1   module My_submodule = struct
2     let f_sub x =
3       let fourty_two = 42 in
4       let int_const = fourty_two + fourty_two in
5       let boxed_const = int_const, int_const in
6       x + int_const * (fst boxed_const)
7     [@@inline never]
8   end
9   
10  let f x =
(gdb) bt
#0  Small.My_submodule.f_sub (x=42 : int) at small.ml:2
#1  0x000000000041ada0 in Small () at small.ml:17
#2  0x0000000000418579 in caml_program ()
(gdb) info locals
boxed_const = (84, 84) : int * int
fourty_two = 42 : int
int_const = 84 : int
(gdb) disas
Dump of assembler code for function Small.My_submodule.f_sub:
=> 0x000000000041ad70 <+0>: add    $0x3720,%rax
   0x000000000041ad76 <+6>: retq   
   0x000000000041ad77 <+7>: nop
End of assembler dump.
```

**Comment by lijunsong** [comment]
@mshinwell @xavierleroy Thanks. Sorry about the intrusive question about SPARC on this thread. Anyway, I have access to SPARC machines, so I can help test on SPARC/Solaris. I'll try to compile and report bugs on Mantis.

File: middle_end/base_types/compilation_unit.ml [hunk 1]
```diff
@@ -71,3 +72,13 @@ let get_current_exn () =
...
   | None -> Misc.fatal_error "Compilation_unit.get_current_exn"
 let get_current_id_exn () = get_persistent_ident (get_current_exn ())
+
+let startup =
+  lazy ({
+    id = Ident.create_persistent "*startup*";
+    linkage_name = Linkage_name.create "caml_startup";
+    hash = Hashtbl.hash "*startup*";
+    is_startup = true;
+  })
```
Comment (chambart) [inline middle_end/base_types/compilation_unit.ml]: Note that Ident.create_persistent does not increment ident counter, so it does not need to be delayed by a lazy.

File: middle_end/base_types/compilation_unit.ml [hunk 2]
```diff
@@ -20,6 +20,7 @@ type t = {
   id : Ident.t;
   linkage_name : Linkage_name.t;
   hash : int;
+  is_startup : bool;
```
Comment (chambart) [inline middle_end/base_types/compilation_unit.ml]: This field is not really used. What is the motivation for this ?

**Comment by mshinwell** [comment]
@chambart has kindly reviewed `Available_regs`, and I shall be addressing his comments in due course.

**Comment by mshinwell** [comment]
Some more highlights in the latest version:
- Almost entirely rewritten Flambda support for propagating information about bindings that have been optimized away, using "phantom let".  This construct effectively has static but no dynamic semantics, but is wholly within the Flambda language.  The change in type of the defining expressions of `let`-expressions forces one to decide what to do in each case depending on whether the `let` is "normal" or "phantom".
- New `Free_names` module in Flambda for holding sets of free variables/symbols and various improvements to the computation of such.  This abstraction should help in future in any case.
- Ability to propagate descriptions of closure and free variable moves/projections all the way through to DWARF.
- Changed `Reg` to help track names properly.  There is now a two-level hierarchy: values of type `Reg.t` may share an underlying register (defined using the record as in previous versions) but yet have different names.  This enables propagation of names even in tricky cases such as operations that clobber their argument (e.g. an in-place multiplication).
- Various bugs fixed, although this is ongoing.

**Comment by mshinwell** [comment]
I forgot one other notable change: Cmm variables are now marked as immutable or mutable (and the vast majority are immutable).  The reason for this is to improve name propagation on registers.

This branch will have some minor code generation differences from the current compiler, but I am going to check that they are indeed minor.

**Comment by mshinwell** [comment]
Also: one thing that isn't here yet but should be coming soon is to use a different type (other than [Ident]) for tracking of identifier names through the backend.  The reason is that we need to be able to express that a given set of identifiers actually correspond to the same original source-level name.  (Consider inlining a function twice inside another function, for example.)  I think this should be a fairly minor change.

**Comment by mshinwell** [comment]
I've been wondering about how to change the command line flags for generation of debugging information (see also http://caml.inria.fr/mantis/view.php?id=6238).  Perhaps something like:
- `-gbacktraces`: stash backtraces upon exception raises
- `-gbounds`: add multiple call sites for `caml_ml_array_bound_error`
- `-gdwarf-line`: emit DWARF source line number information
- `-gdwarf`: full DWARF debugging information (implies `-gdwarf-line`), as provided by this pull request.  This is separate from `-gdwarf-line` as it is a major and intricate addition to the compiler that will probably take one or two releases to fully stabilize
- `-gsymbols`: keep symbols so they are visible in the debugger.

`-g` would select what is currently considered the stable and reasonable subset of the debugging options.  To start with this would be `-gbacktraces`, `-gbounds`, `-gdwarf-line` and `-gsymbols`, matching the existing compiler.

File: middle_end/free_names.ml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    t.free_phantom_variables
+      <- Variable.Set.union t.free_phantom_variables t'.free_phantom_variables;
+    union_free_symbols_only t t'
+
+  let bound_variables t bound_vars =
+    t.free_variables <- Variable.Set.diff t.free_variables bound_vars;
+    t.free_phantom_variables
+      <- Variable.Set.diff t.free_phantom_variables bound_vars
+
+  let freeze t = t
```
Comment (chambart) [inline middle_end/free_names.ml]: This should probably be:

``` OCaml
let freeze t = { t with free_variables = t.free_variables }
```

To do a copy.

File: middle_end/flambda.mli [hunk 4]
```diff
@@ -304,14 +358,13 @@ and function_declaration = private {
   body : t;
   (* CR-soon mshinwell: inconsistent naming free_variables/free_vars here and
      above *)
```
Comment (chambart) [inline middle_end/flambda.mli]: This CR is apparently fixed by the change to free_names

File: middle_end/flambda.mli [hunk 5]
```diff
@@ -372,14 +425,36 @@ module Constant_defining_value :
 
 type expr = t
 
+type symbol_provenance = {
+  names : string list;
```
Comment (chambart) [inline middle_end/flambda.mli]: This field probably need a comment: where is the information coming from ?

**Comment by mshinwell** [comment]
New target start of 4.05 release cycle

File: middle_end/flambda.mli (hunk 4 above)
Comment (mshinwell) [inline middle_end/flambda.mli]: Removed

File: middle_end/free_names.ml (hunk 3 above)
Comment (mshinwell) [inline middle_end/free_names.ml]: We should copy all of the members, no?  I've made that change.

File: middle_end/flambda.mli (hunk 5 above)
Comment (mshinwell) [inline middle_end/flambda.mli]: Yeah, I've added a comment.

**Comment by DemiMarie** [comment]
I realized a dealbreaker for using this with GDB: licensing.

OCaml's license (LGPLv2.1 with exceptions) is incompatible with GDB's license (GPLv3+).  So a plugin for GDB that used any part of OCaml could not be lawfully distributed.

The only solutions I know of are to either relicense the OCaml distribution as LGPLv2.1+ with exceptions, or to drop the idea of GDB support and use LLDB exclusively.

**Comment by johanatan** [comment]
@DemiMarie there is a third option: compile the OCaml bits into a binary and do marshaling/IPC across module boundary (even a shared dynlib loaded in the same process space should be fine).

**Comment by DemiMarie** [comment]
@johanatan Debian and the FSF disagree with you when it comes to shared libraries (although the Linux kernel devs have tolerated proprietary modules).  But using IPC should be fine.

If you use IPC, **be sure that it is secure**.  stdin/stdout or an AF_UNIX socket in the user's home directory is fine.  An AF_UNIX socket in a subdirectory of `/tmp` is fine if you are willing to leak the directory in `/tmp`.  Anything else is likely to require cryptographic authentication of one or both parties.

**Comment by johanatan** [comment]
I don't mind Debian and the FSF disagreeing with me. Any legitimate judge (such as William Haskell Alsup) would easily see through their folly.

But, sure, if you are the squeamish type, stick to IPC.

**Comment by mshinwell** [comment]
@DemiMarie The matters of licensing have been thought about.  As far as I am aware, the LGPLv2.1 licence under which OCaml is now distributed is compatible with the GPLv3 licence of gdb.  See https://www.gnu.org/licenses/license-list.html.  Can you explain why you think that is not the case?

**Comment by DemiMarie** [comment]
The GPL is only compatible with the same version of the LGPL.  So the
LGPLv2.1 is compatible with the GPLv2, but not the GPLv3.  The OCaml
linking exception doesn't change anything here.

**Comment by lpw25** [comment]
Not according to the link Mark gave:

> GNU Lesser General Public License (LGPL) version 2.1 (#LGPLv2.1)
> This is the previous version of the LGPL: a free software license, but not a strong copyleft license, because it permits linking with nonfree modules. It is compatible with GPLv2 and GPLv3. We generally recommend the latest version of the LGPL, for special circumstances only. To learn more about how LGPLv2.1 is compatible with other GNU licenses, please see our FAQ.

**Comment by DemiMarie** [comment]
Then you are correct.  I mixed this up with the LGPLv3 and the GPLv2 which
I believe are incompatible.

**Comment by ghost** [comment]
> @DemiMarie The matters of licensing have been thought about. As far as I am aware, the LGPLv2.1 licence under which OCaml is now distributed is compatible with the GPLv3 licence of gdb. See https://www.gnu.org/licenses/license-list.html. Can you explain why you think that is not the case?

I'm not a license expert but my reading of the fine print is that it's not as simple as that.

If you look at the [compatibility matrix](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility), the combinations which would seem to apply suggest that [item 7](https://www.gnu.org/licenses/gpl-faq.html#compat-matrix-footnote-7) must be considered. Notably: _"If you can switch the LGPLed code in this case to using an appropriate version of the GPL instead (as noted in the table), you can make this combination."_

As far as I understand, the relicensing of LGPLv2.1 code under later versions is only trivial if the optional ["any later version"](https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.en.html#VersionTwoOrLater) clause is used. But if you look at the current OCaml [LICENSE](https://github.com/ocaml/ocaml/blob/trunk/LICENSE#L5) it does not seem to use [that clause](https://github.com/ocaml/ocaml/blob/trunk/LICENSE#L179).

**Comment by jordwalke** [comment]
I've mentioned this progress to other OCamlers at Facebook, and we are all very excited about this effort. How far off is this diff from being accepted?

**Comment by mshinwell** [comment]
@freebroccolo I do not believe item 7 applies.  The situation here involves three parts combined into a program:
1. gdb (GPLv3)
2. libmonda (MIT)
3. compiler-libs and the OCaml runtime (LGPLv2.1)
gdb links dynamically to the libmonda shared object which is statically linked to compiler-libs and the runtime.  There is no copying of any code from the OCaml compiler distribution into gdb.  So the combination is I believe "using a library" (in the context of the compatibility matrix you link to) which does not have any item 7 restriction.

**Comment by mshinwell** [comment]
@jordwalke I am preparing some alpha/beta-ish OPAM packages today to get some more feedback on this.  I believe we will be in a state to merge into 4.05.  There will still be a fair amount of potential future work to do, but the initial patch should still be substantively useful.

**Comment by ghost** [comment]
@mshinwell okay, this is good to know then, thanks for clarifying. I'm also really looking forward to this and was just a little worried when I saw the potential licensing issue come up.

**Comment by DemiMarie** [comment]
I think that an LLDB plugin is better, precisely because of licensing.

**Comment by DemiMarie** [comment]
If OCaml didn't have its linking exception, this wouldn't work.  As is, not
sure.

**Comment by gasche** [comment]
I would like this PR to focus on the technical and design aspects of Mark's work, which are modifications to the OCaml compiler distribution to improve native debugger support. I think that discussions of fine points of licensing, or tool preferences (GDB vs. LLDB), are best discussed outside, to keep the current thread on topic and encourage people to review the proposed changes. Some of the modifications are somewhat invasive and any external help in reviewing, understanding and questioning the changes is definitely helpful and warmly welcome.

**Comment by mshinwell** [comment]
For those who dare...
- Use the `4.05.0+monda` switch from `github.com/mshinwell/opam-repo-dev`
- `opam install libmonda-gdb`
- Compile your OCaml code with `-g`.  Compilation may sometimes be slow.  Flambda is enabled.
- Debug with `gdb-ocaml` not `gdb`.  You can also use this to attach to an existing process.

In the debugger:
- `bt` for backtrace (inlined frames are not shown)
- `break` (and associated breakpoint commands), `run`, `continue`, `step`, `next`, `finish` should work (but return values are not printed correctly when using `finish`)
- `info args` will show arguments
- `info loc` will show locals.  This shows _all_ locals from the current function at present, even if currently out of scope.  Variables that do not have unique names are disambiguated using their locations.
- `p` should work for printing variable names although I think there may be problems trying to get it to print variables with locations in their names (use `info args` or `info loc` instead).
- `p` works for toplevel constant and inconstant values e.g. `p Mymodule.foo` (except for non-allocated constants, such as integers, which are not currently visible).  Tab completion won't work on these names.
- `p` understands record paths e.g. `foo.label1.label2`.  It should understand array indexing notation (also on lists) but this needs a bit more work
- `disas` will show disassembly annotated correctly with OCaml function names
- `apropos ocaml` will reveal various settings e.g. for the maximum value printing depth and the length after which strings are truncated.

Tab completion can be tricky: it doesn't usually work if you type "Foo.<tab>".  Instead complete before or after writing the dot.

Conventions:
- For nested functions: `Foo.bar.baz` means the function called `baz` defined (as in, statically scoped) inside the function `Foo.bar`.  Anonymous functions are named based on their source location.
- For module initialisers: `Foo` is the name of the module initialiser for the module of the same name.  Breakpoints may be set on these functions.

Some kinds of values are not yet printed properly.  Boxed polymorphic variant constructions are one such, together with values with complicated types such as certain functor constructions.

**Comment by mshinwell** [comment]
Some more points if you want to try this:
- You may need a "binutils-dev" package or similar (if you lack `/usr/include/bfd.h`).
- You probably want `OPAMKEEPBUILDDIR=true` so that `.cmt` files can be found.
- This probably isn't building on OS X at the moment.

**Comment by kayceesrk** [comment]
Here is a handy dockerfile for testing: https://gist.github.com/kayceesrk/94cdd19931b324a247bbd1a58ad47b47

**Comment by mshinwell** [comment]
I've started a page with some information on this project here:
http://mshinwell.github.io/libmonda/

**Comment by DemiMarie** [comment]
Status?

**Comment by mshinwell** [comment]
I'm working on #756, which has turned into a bit of an epic, to unblock this.  I expect we will get everything merged in time for 4.06.

**Comment by bikallem** [comment]
Status - Will this be ready for 4.07? If not could this be prioritised for 4.07 please? This would save countless hours of printf/format debugging in ocaml. 

Question - will this support 'dynload' debugging too?

Also I am ready to monkey/user-test this feature once it is ready and/or if anyone here needs the help.

**Comment by mshinwell** [comment]
I hope this will be in for 4.07.

**Comment by bikalgurung** [comment]
@mshinwell we are making a "fuss" about this GPR in discuss. (https://discuss.ocaml.org/t/how-do-you-debug-ocaml/981/15)

**Comment by XVilka** [comment]
Seems like needs to be rebased

**Comment by mshinwell** [comment]
This pull request provides a pass that constructs DWARF information from terms in the Linearize language and then emits it to the assembly output.  It uses the register availability and phantom let information passed through from earlier in the backend.  This patch also provides the new command-line flags for enhanced DWARF information generation together with setting of the DWARF output format and standard level.

@bschommer This might give you some more context for the `Dwarf_high` library.  (Just read any changeset which starts "GPR#574"; currently there are only two.)  I need to re-test this as I've had to alter various things, but this is pretty close to a final form.

One thing that may still need some thought is what the final set of "-g" options should look like (see my older comment above about that).

**Comment by mshinwell** [comment]
(We should make sure the commit message for this pull request describes what is added.  cc @trefis )

**Pull request closed without merging**